runtime/src/iree/hal/drivers/hip/hip_device.c

```diff
@@ -1110,31 +954,36 @@ static iree_status_t iree_hal_hip_device_queue_dealloca(
     const iree_hal_semaphore_list_t wait_semaphore_list,
     const iree_hal_semaphore_list_t signal_semaphore_list,
     iree_hal_buffer_t* buffer) {
-  iree_hal_hip_device_t* device = iree_hal_hip_device_cast(base_device);
-  IREE_RETURN_IF_ERROR(
-      iree_hal_hip_set_context(device->hip_symbols, device->hip_context));
+  IREE_TRACE_ZONE_BEGIN(z0);
 
-  if (iree_hal_hip_allocator_isa(iree_hal_device_allocator(base_device))) {
-    iree_status_t status = iree_hal_deferred_work_queue_enqueue_dealloc(
-        device->work_queue, wait_semaphore_list, signal_semaphore_list, buffer);
-    if (iree_status_is_ok(status)) {
-      status = iree_hal_deferred_work_queue_issue(device->work_queue);
-    }
-    return status;
-  }
+  iree_hal_hip_device_t* device = iree_hal_hip_device_cast(base_device);
 
   // NOTE: block on the semaphores here; we could avoid this by properly
   // sequencing device work with semaphores. The HIP HAL is not currently
   // asynchronous.
-  IREE_RETURN_IF_ERROR(iree_hal_semaphore_list_wait(wait_semaphore_list,
-                                                    iree_infinite_timeout()));
+  IREE_RETURN_AND_END_ZONE_IF_ERROR(
+      z0, iree_hal_semaphore_list_wait(wait_semaphore_list,
+                                       iree_infinite_timeout()));
 
   // Schedule the buffer deallocation if we got it from a pool and otherwise
   // drop it on the floor and let it be freed when the buffer is released.
   iree_status_t status = iree_ok_status();
   if (device->supports_memory_pools) {
+    uint64_t device_num = 0;
+    if (queue_affinity) {
```

Comment:
don't wait and _then_ error on an input argument - check all preconditions before you make any state changes unless it's very costly